src/picongpu/include/particles/flylite/NonLTE.tpp

```diff
@@ -0,0 +1,206 @@
+                    )
+                )
+            );
+
+        if( ! dc.hasId( helperFields::LocalDensity::getName( "electrons" ) ) )
+            dc.share(
+                std::shared_ptr< ISimulationData >(
+                    new helperFields::LocalDensity(
+                        "electrons",
+                        m_avgGridSizeLocal
+                    )
+                )
+            );
+
+        /* for each ion species */
+        if( ! dc.hasId( helperFields::LocalRateMatrix::getName( ionSpeciesName ) ) )
+            dc.share(
+                std::shared_ptr< ISimulationData >(
+                    new helperFields::LocalRateMatrix(
+                        ionSpeciesName,
+                        m_avgGridSizeLocal
+                    )
+                )
+            );
+
+        if( ! dc.hasId( helperFields::LocalDensity::getName( ionSpeciesName ) ) )
+            dc.share(
+                std::shared_ptr< ISimulationData >(
+                    new helperFields::LocalDensity(
+                        ionSpeciesName,
+                        m_avgGridSizeLocal
+                    )
+                )
+            );
+    }
+
+    template<
+        //! @todo for multi ion species IPD: typename T_OtherIonsList,
+
+        typename T_ElectronsList,
+        typename T_PhotonsList
+    >
+    template<
+        typename T_IonSpeciesType
+    >
+    void
+    NonLTE<
+        //! @todo for multi ion species IPD: T_OtherIonsList,
+
+        T_ElectronsList,
+        T_PhotonsList
+    >::update(
+        std::string const ionSpeciesName,
+        uint32_t currentStep
+    )
+    {
+        using IonSpeciesType = T_IonSpeciesType;
+
+        // calculate density fields and energy histograms
+        fillHelpers< IonSpeciesType >( ionSpeciesName, currentStep );
+
+        //! @todo calculate rate matrix
+        //! @todo implicit ODE solve to evolve populations
+        //! @todo modify f_e of free electrons
+        //! @todo modify f_ph of photon field (absorb)
+        //! @todo change charges, create electrons & photons
+    }
+
+    template<
+        //! @todo for multi ion species IPD: typename T_OtherIonsList,
+
+        typename T_ElectronsList,
+        typename T_PhotonsList
+    >
+    template<
+        typename T_IonSpeciesType
+    >
+    void
+    NonLTE<
+        //! @todo for multi ion species IPD: T_OtherIonsList,
+
+        T_ElectronsList,
+        T_PhotonsList
+    >::fillHelpers(
+        std::string const ionSpeciesName,
+        uint32_t currentStep
+    )
+    {
+        using IonSpeciesType = T_IonSpeciesType;
+
+        // calculate density fields
+        helperFields::FillLocalDensity< MakeSeq_t< IonSpeciesType > > fillDensityIons{};
+        fillDensityIons(
+            currentStep,
+            ionSpeciesName
+        );
+
+        helperFields::FillLocalDensity< T_ElectronsList > fillDensityElectrons{};
+        fillDensityElectrons(
+            currentStep,
+            "electrons"
```

Comment:
because for collisional-radiative modeling, physically only actual *electrons* and actual *photons* exist as elementary particle that are relevant for the process.

I group those together and allow the user to specify a list of particles in `flylite.param` which makes up all each of the two groups.